### PR TITLE
implement `incbench` command for ease-of-use benchmark

### PR DESCRIPTION
**Pull request by xin3he**
## Type of Change

feature

## Description

- [x] implement `incbench` command as entrypoint for ease-of-use benchmark
- [x] automatically check numa/socket info and dump it with table for ease-of-understand
- [x] supports both Linux and Windows platform
- [x] add benchmark documents
- [x] dump benchmark summary
- [x] add benchmark UTs

### General Use Cases

1. `incbench main.py`: run 1 instance on NUMA:0.
2. `incbench --num_i 2 main.py`: run 2 instances on NUMA:0.
3. `incbench --num_c 2 main.py`: run multi-instances with 2 cores per instance on NUMA:0.
4. `incbench -C 24-47 main.py`: run 1 instance on COREs:24-47.
5. `incbench -C 24-47 --num_c 4 main.py`: run multi-instances with 4 COREs per instance on COREs:24-47.

> Note:
    > - `num_i` works the same as `num_instances`
    > - `num_c` works the same as `num_cores_per_instance`

## Expected Behavior & Potential Risk

image

## How has this PR been tested?

tested on Ultra7 1003H (1 socket, 6 P-Cores, 8 E-Cores and 2 SoC E-Cores) and cascade lake(2 numa, 24 cores per numa)

## Dependency Change?

N/A

### CODE REVIEW

**Comment by xin3he** [comment]
demo: Multiple Instance Benchmark Summary
image

File: neural_compressor/common/benchmark.py [hunk 1]
```diff
@@ -0,0 +1,528 @@
...
+        args (argparse): arguments for setting different configurations
+        core_list_per_instance (dict): {"instance_index": ["node_index", "cpu_index", num_cpu]}
+        raw_cmd (str): script.py and parameters for this script
+    """
+    instance_cmd = ""
+    if not os.getenv("PYTHON_PATH"):  # pragma: no cover
+        logger.info("The interpreter path is not set, using string `python` as command.")
+        logger.info("To replace it, use `export PYTHON_PATH=xxx`.")
+    interpreter = os.getenv("PYTHON_PATH", "python")
+    current_work_dir = os.getcwd()
```
Comment (yiliu30) [inline neural_compressor/common/benchmark.py:380]: Suggest reuse [DEFAULT_WORKSPACE](https://github.com/intel/neural-compressor/blob/4372a762585189accc65196e081a0a7a85f5af9e/neural_compressor/common/utils/constants.py#L45).

File: neural_compressor/common/benchmark.py [hunk 2]
```diff
@@ -0,0 +1,528 @@
...
+    """Collect numa/socket information on linux system.
+
+    Returns:
+        numa_info (dict):   demo: {numa_index: {"physical_cpus": "xxx"; "logical_cpus": "xxx"}}
+                            E.g.    numa_info = {
+                                        0: {"physical_cpus": "0-23", "logical_cpus": "0-23,48-71"},
+                                        1: {"physical_cpus": "24-47", "logical_cpus": "24-47,72-95"}
+                                    }
+    """
+    result = subprocess.run(["lscpu"], capture_output=True, text=True)
```
Comment (yiliu30) [inline neural_compressor/common/benchmark.py:62]: Consider adding a log to guide the user if `lscpu` or `numactl` fails or is not found.

File: neural_compressor/common/benchmark.py [hunk 3]
```diff
@@ -0,0 +1,528 @@
...
+        physical_cpus = psutil.cpu_count(logical=False)
+        logical_cpus = psutil.cpu_count(logical=True)
+        physical_cpus_per_socket = physical_cpus // num_socket
+        logical_cpus_per_socket = logical_cpus // num_socket
+        for i in range(num_socket):
+            physical_cpus_str = str(i * physical_cpus_per_socket) + "-" + str((i + 1) * physical_cpus_per_socket - 1)
+            if num_socket == 1:
+                logical_cpus_str = str(i * logical_cpus_per_socket) + "-" + str((i + 1) * logical_cpus_per_socket - 1)
+            else:
+                remain_cpus = logical_cpus_per_socket - physical_cpus_per_socket
```
Comment (yiliu30) [inline neural_compressor/common/benchmark.py:94]: nit: Looks like these codes are redundant in  `get_linux_numa_info` and `get_windows_numa_info`.

File: neural_compressor/common/benchmark.py [hunk 4]
```diff
@@ -0,0 +1,528 @@
+# Copyright (c) 2024 Intel Corporation
```
Comment (yiliu30) [inline neural_compressor/common/benchmark.py:1]: Please add some UTs.

**Review by yiliu30 — APPROVED** [review]
LTGM!

File: neural_compressor/common/benchmark.py (hunk 2 above)
Comment (xin3he) [reply neural_compressor/common/benchmark.py:62]: I think shell can tell user this info.

File: neural_compressor/common/benchmark.py (hunk 3 above)
Comment (xin3he) [reply neural_compressor/common/benchmark.py:94]: The logic of the two functions is different, and merging them will make the code unclear. Reusing the common logic will introduce a lot of parameters in the new function, so I prefer to keep it.

File: neural_compressor/common/benchmark.py (hunk 4 above)
Comment (xin3he) [reply neural_compressor/common/benchmark.py:1]: Good point, I will.

File: examples/3.x_api/pytorch/nlp/huggingface_models/language-modeling/quantization/smooth_quant/run_clm_no_trainer.py [hunk 5]
```diff
@@ -233,32 +235,25 @@ def run_fn(model):
         else:
             acc = results["results"][task_name]["acc,none"]
     print("Accuracy: %.5f" % acc)
-    print('Batch size = %d' % args.batch_size)
+    print("Batch size = %d" % args.batch_size)
 
 if args.performance:
     user_model.eval()
-    from intel_extension_for_transformers.transformers.llm.evaluation.lm_eval import evaluate, LMEvalParser
+    batch_size, input_leng = args.batch_size, 512
```
Comment (thuang6) [inline examples/3.x_api/pytorch/nlp/huggingface_models/language-modeling/quantization/smooth_quant/run_clm_no_trainer.py:242]: looks like we decide to change the design to skip using ITREX lm_eval and no reporting accuracy.  was it agreed in design review?

Comment (xin3he) [reply examples/3.x_api/pytorch/nlp/huggingface_models/language-modeling/quantization/smooth_quant/run_clm_no_trainer.py:242]: It's a performance test so I don't think we need to use real data.
BTW, we have decided to re-write this example per Chang's request. This change will be delete and currently it's for our extension test. https://jira.devtools.intel.com/browse/ILITV-3682

**Review by thuang6 — APPROVED** [review]
LGTM

**Comment by xin3he** [comment]
ext test: https://inteltf-jenk.sh.intel.com/job/intel-lpot-validation-top-pr-extension-3x-API/60/

**Comment by xin3he** [comment]
The final demo:
image

**Comment by xin3he** [comment]
example test: https://inteltf-jenk.sh.intel.com/job/intel-lpot-validation-top-pr-extension-3x-API/66/
